### DISCUSSION TITLE
Bridge contract parameters

### DISCUSSION THREAD

**Discussion by enochSatis** [zkSync-Community-Hub/zksync-developers#79] [General] [answered]
### Environment

Testnet

### Provide a brief description of the functionality you're trying to implement.

I am currently trying to integrate zkSync's bridge, particularly the ERC20 bridge. For security reasons, I would like to deposit by calling bridge contracts directly instead of using any SDKs. 

### What is the specific issue or error you're encountering?

My deposit transactions sometimes got reverted. I am just guessing the transaction's `value` field, as I have no clue about how it is computed. 

Considering the deposit function in L1 ERC20 bridge contract:
```
deposit( address _l2Receiver, address _l1Token, uint256 _amount, uint256 _l2TxGasLimit, uint256 _l2TxGasPerPubdataByte )
```

What I know:
- `_l2Receiver`, `_l1Token` and `_amount` are intuitive
- `_l2TxGasPerPubdataByte` is 800, and is a fixed parameter according to the docs (https://era.zksync.io/docs/api/js/utils.html#required-l1-to-l2-gas-per-pubdata-limit)

What I am still guessing:
- `_l2TxGasLimit` should be just the gas limit for L2 transaction (release tokens to receiver)
- The `value` of ETH transferred to the bridge contract should be paying this gas fee
- The transferred ETH should be somewhat proportional to this `_l2TxGasLimit`

My doubts:
1. How could I compute the required transferred ETH amount to the bridge?
2. Is there a good estimation for `_l2TxGasLimit` instead of hardcoding?

### Can you share the error messages or logs you're receiving, if any?

`Fail with error 'mv'`

ref: https://goerli.etherscan.io/tx/0xae649b09ec2860ca0fd07af5cf14a69b1767a4c1aaa8e3415a1d7f71cf4a0f34

---

Transaction that success

https://goerli.etherscan.io/tx/0x0a93499f9326ce01c6587a25e68a773b7177cb16d1050784e0d7d57716ce1e22

---

They differ by the transferred ETH value (by a factor of 10).

### Have you made any recent changes to the contract before encountering this issue?

na

### Are there any external libraries or contracts that your contract interacts with?

na

### Can you provide the relevant portions of your contract code where the issue is occurring?

na

### Have you tried to isolate the problem, and if so, what were the results?

na

### What steps have you already taken to try to resolve the issue?

na

**Comment by enochSatis** [comment]
Update:

After reading through source codes and docs, I have better understandings:

## "mv" error

In the `mailbox.sol` contract:

```
require(msg.value >= baseCost + _l2Value, "mv");
```

This denoted that the transaction value must be larger than the L2 base cost  and L2 value. For deposit action in the official bridge, it seems that the `_l2Value` in the L2 release deposit fund part is 0. Hence, the transferred value should be only able to cover the L2 base cost, which is:

```
L2 gas price * L2 TX gas limit
```

## "my" error

```
require(_totalGasLimit >= overhead, "my");
```

I guess this means the `_l2TxGasLimit` must be at least equal to the overhead gas (gas required for writing L1 proof?). That explains why I would get this error if I input a small `_l2TxGasLimit`.

## Trials to estimate L2 TX Gas Limit

Still, I have no idea how I could compute the `_l2TxGasLimit`. 

I followed the docs and call the RPC API method of `zks_estimateGasL1ToL2` to use the result as `_l2TxGasLimit` (https://era.zksync.io/docs/api/api.html#zks-estimategasl1tol2). I tried putting the L1 and L2 part of the deposit transaction into the `FROM`, `TO` and `DATA` field. 

```
curl -X POST -H "Content-Type: application/json" --data '{"jsonrpc":"2.0", "id":2, "method": "zks_estimateGasL1ToL2", "params": [ { "from": ${FROM}, "to":${TO}, "data": ${DATA} } ] }' "https://testnet.era.zksync.dev"
```

## Unmatched values for L2 TX Gas Limit

The `zks_estimateGasL1ToL2` method returns hex `0x3a84859` no matter what, which is `61360217` in dec. 

When trancing deposit transactions through Goerli etherscan, however, shows the `_l2TxGasLimit` is only `11922e` in hex, `1151534` in dec.

The values differ by 60 times, which makes no sense. 

**So which is the correct way to compute the `_l2TxGasLimit` for L1 bridge `deposit` function?**

**Comment by vladbochok** [comment]
Hello @enochSatis,

Firstly, I want to pay attention to `_l2TxGasPerPubdataByte` parameter

> _l2TxGasPerPubdataByte is 800, and is a fixed parameter according to the docs

Even though `_l2TxGasPerPubdataByte` is constant for now, it may (and will) change in the future, so please don't rely on the exact value.

Regarding L1 -> L2 transaction gas cost estimation, indeed the correct way to estimate gas would be using  `zks_estimateGasL1ToL2` API method. Please note, gas cost on L2 **depends on the L1 gas price**, so the resulting value can be smaller or bigger. I think that's the reason why you see some transactions with different L2 gas limits.

**Reply by enochSatis** [reply]
Thanks for the reply @vladbochok .

## _l2TxGasPerPubdataByte

When I trace the `Mailbox.sol` contract, I see:

```
require(_l2GasPerPubdataByteLimit == REQUIRED_L2_GAS_PRICE_PER_PUBDATA, "qp");
```

Seems that this value is a modifiable number in the contract. I don't see any RPC API method to get this number either. Apart from using a variable instead of a hardcoded constant in my code, is there any external method (eg. RPC API `zks_xxx` method, or SDK) I can get in case it changes in the future?

## _l2TxGasLimit

Sorry I still don't quite understand the `zks_estimateGasL1ToL2` method. I tried estimating these 2 TXs on testnet:

```
curl -X POST -H "Content-Type: application/json" \
--data '{"jsonrpc":"2.0", "id":2, "method": "zks_estimateGasL1ToL2", "params": [ { "from": ${L1_sender}, "to":${L1_bridge}, "data": ${DATA} } ] }' \
"https://testnet.era.zksync.dev"

curl -X POST -H "Content-Type: application/json" \
--data '{"jsonrpc":"2.0", "id":2, "method": "zks_estimateGasL1ToL2", "params": [ { "from": ${L2_bridge}, "to":${L2_receiver}, "data": ${DATA} } ] }' \
"https://testnet.era.zksync.dev"
```

but they both returned 61360217 gas, which is 60 times larger than the contract input for `_l2TxGasLimit` shown in Goerli etherscan. I do understand L2 gas is dependent on L1 gas price, but nearly all deposit transactions on Goerli showed `_l2TxGasLimit = ~1.1e6`, before and after I called the `zks_estimateGasL1ToL2` method. **Is this API always returning an overestimated number to guarantee the L1 to L2 transactions will pass?**

For 60 times larger `_l2TxGasLimit`, I would need to transfer 60 times more ETH to the L1 bridge contract. I do understand that the unused gas will be refunded, but transferring 60 times more ETH as TX value is quite a lot of liquidity, and is not feasible in production Mainnet.

Thanks for your help~

**Reply by StanislavBreadless** [reply]
@enochSatis could you please provide the exact from/to/data you are using?

**Reply by enochSatis** [reply]
@StanislavBreadless 

Here is my curl method:

```
curl -X POST -H "Content-Type: application/json" --data '{"jsonrpc":"2.0", "id":2, "method": "zks_estimateGasL1ToL2", "params": [ { "from": "0xc4adcF8814a1da13522716A23331Ce4d48A1414d", "to":"0x927DdFcc55164a59E0F33918D13a2D559bC10ce7", "data": "0x933999fb000000000000000000000000e33e60aaa7d479675933dcfd0868addf49efeb3d000000000000000000000000d35cceead182dcee0f148ebac9447da2c4d449c400000000000000000000000000000000000000000000000000000000000f4240000000000000000000000000000000000000000000000000000000000011da500000000000000000000000000000000000000000000000000000000000000320" } ] }' "https://testnet.era.zksync.dev"
```

zks_estimateGasL1ToL2 API RES:
```
{"jsonrpc":"2.0","result":"0x3a84859","id":2}
```

My raw TX:
```
{'chainId': 5, 'gas': 1000000, 'gasPrice': 7234, 'nonce': 21448, 'value': 300000000000000, 'to': '0x927DdFcc55164a59E0F33918D13a2D559bC10ce7', 'data': '0x933999fb000000000000000000000000e33e60aaa7d479675933dcfd0868addf49efeb3d000000000000000000000000d35cceead182dcee0f148ebac9447da2c4d449c400000000000000000000000000000000000000000000000000000000000f4240000000000000000000000000000000000000000000000000000000000011da500000000000000000000000000000000000000000000000000000000000000320'}
```

Thanks.

**Reply by StanislavBreadless** [reply, marked as answer]
@enochSatis you need to explicitly provide the gas per pubdata that you will use in case of L1->L2 transactions. For L1->L2 transactions is 800, i.e. 320 in hex:

```
curl -X POST -H "Content-Type: application/json" --data '{"jsonrpc":"2.0", "id":2, "method": "zks_estimateGasL1ToL2", "params": [ { "from": "0xc4adcF8814a1da13522716A23331Ce4d48A1414d", "to":"0x927DdFcc55164a59E0F33918D13a2D559bC10ce7", "data": "0x933999fb000000000000000000000000e33e60aaa7d479675933dcfd0868addf49efeb3d000000000000000000000000d35cceead182dcee0f148ebac9447da2c4d449c400000000000000000000000000000000000000000000000000000000000f4240000000000000000000000000000000000000000000000000000000000011da500000000000000000000000000000000000000000000000000000000000000320", "eip712Meta": { "gasPerPubdata": "320" } } ] }' "https://testnet.era.zksync.dev"
```

Response is 0xce8c9, which is more realistic

**Reply by enochSatis** [reply]
I see. Thanks a lot @StanislavBreadless !

**Comment by guarinojoe** [comment]
To compute the required transferred ETH amount to the bridge for depositing ERC20 tokens, you need to estimate the gas cost for the L2 transaction and include a buffer for safety. You can use gas estimation tools or monitor gas prices on the network for a more accurate estimation. The _l2TxGasLimit parameter should indeed reflect the gas limit for the L2 transaction, and the transferred ETH should cover this gas cost. It's advisable to adjust the transferred ETH amount based on network conditions and gas prices to ensure successful transactions. Additionally, consider implementing gas price strategies to dynamically adjust the transferred ETH amount based on current gas prices.

**Comment by beckyhaynes** [comment]
To compute the required transferred ETH amount for depositing to the zkSync ERC20 bridge contract, you need to estimate the gas cost for the L2 transaction (release tokens to receiver) and add an additional buffer for safety. Here's a general approach:

Estimate Gas Cost: You can estimate the gas cost for the L2 transaction by simulating the transaction locally or using historical data from similar transactions. Take into account the gas price and limit parameters.

Add Buffer: Add an additional buffer to the estimated gas cost to ensure the transaction has enough ETH to cover any unforeseen circumstances or fluctuations in gas prices.

Compute Transferred ETH: The transferred ETH should be the sum of the estimated gas cost and the buffer amount.

Set _l2TxGasLimit: While there's no fixed formula for determining _l2TxGasLimit, you can experiment with different values based on your estimation and observations. Start with a conservative estimate and adjust as needed.

Remember to monitor gas prices and adjust your calculations accordingly to ensure successful transactions.

**Comment by hanjv** [comment]
To compute the required transferred ETH amount for depositing into zkSync's ERC20 bridge contract, you'll need to estimate the gas cost for the L2 transaction and ensure it's covered by the transferred ETH. The gas limit (_l2TxGasLimit) and gas price (_l2TxGasPerPubdataByte) parameters play a crucial role in this computation. Consider dynamic gas estimation and adju

**Comment by maestaspat** [comment]
💰 When interacting with zkSync's bridge contract directly, ensure the value transferred with the deposit call covers the gas fee for the L2 transaction.

 To compute the required transferred ETH amount:
    - Estimate the gas cost for the L2 transaction using a reasonable gas price.
    - Add an additional buffer for safety.

 Estimating _l2TxGasLimit:
    - Start with a reasonable default, such as 150,000 gas.
    - Adjust based on network congestion and transaction complexity.

 For the specific issue of transactions getting reverted, ensure that the value transferred covers the gas fee adequately and that the gas limit is set appropriately.

**Comment by sifrediblue** [comment]
The user is trying to integrate zkSync's ERC20 bridge and is encountering issues with deposit transactions sometimes getting reverted. They are guessing at the value field of the deposit function and are unsure how to compute the required transferred ETH amount to the bridge. They are also unsure if there is a good estimation for _l2TxGasLimit instead of hardcoding. The user has provided a failed transaction with the error 'mv' and a successful transaction with a transferred ETH value that differs by a factor of 10. Further investigation and debugging may be needed to determine the cause of the issue and the feasibility and impact of any necessary changes.
